command/connect/redirecttraffic/redirect_traffic.go

```diff
@@ -150,9 +153,25 @@ func (c *cmd) generateConfigFromFlags() (iptables.Config, error) {
 			}
 		}
 
-		svc, _, err := c.client.Agent().Service(c.proxyID, nil)
-		if err != nil {
-			return iptables.Config{}, fmt.Errorf("failed to fetch proxy service from Consul Agent: %s", err)
+		var svc *api.AgentService
+		if c.nodeName == "" {
+			var err error
+			svc, _, err = c.client.Agent().Service(c.proxyID, nil)
+			if err != nil {
+				return iptables.Config{}, fmt.Errorf("failed to fetch proxy service from Consul Agent: %s", err)
+			}
+		} else {
+			svcList, _, err := c.client.Catalog().NodeServiceList(c.nodeName, &api.QueryOptions{
+				Filter:             fmt.Sprintf("ID == %q", c.proxyID),
+				MergeCentralConfig: true,
+			})
+			if err != nil {
+				return iptables.Config{}, fmt.Errorf("failed to fetch proxy service from Consul: %s", err)
+			}
+			if len(svcList.Services) < 1 {
+				return iptables.Config{}, fmt.Errorf("proxy service with ID %q not found", c.proxyID)
+			}
+			svc = svcList.Services[0]
```

Comment:
Nit: We also could check if there is more than 1 service returned (should not happen because of the filter but just in case the filter is broken) - similar to https://github.com/hashicorp/consul/blob/main/command/connect/envoy/envoy.go#L572

Reply:
I originally had that code there, but then removed because I couldn't write a test for it. There should always be one instance with a given service ID on a node. Let me know though if you think it's worth having it just in case.

Reply:
done!

Reply:
Thanks!